presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergePartialAggregationsWithFilter.java

```diff
@@ -0,0 +1,363 @@
+        }
+
+        return plan;
+    }
+
+    private static class Context
+    {
+        private final Map<VariableReferenceExpression, VariableReferenceExpression> partialResultToMask;
+        private final Map<VariableReferenceExpression, VariableReferenceExpression> partialOutputMapping;
+
+        public Context()
+        {
+            partialResultToMask = new HashMap<>();
+            partialOutputMapping = new HashMap<>();
+        }
+
+        public boolean isEmpty()
+        {
+            return partialOutputMapping.isEmpty();
+        }
+
+        public void clear()
+        {
+            partialResultToMask.clear();
+            partialOutputMapping.clear();
+        }
+
+        public Map<VariableReferenceExpression, VariableReferenceExpression> getPartialOutputMapping()
+        {
+            return partialOutputMapping;
+        }
+
+        public Map<VariableReferenceExpression, VariableReferenceExpression> getPartialResultToMask()
+        {
+            return partialResultToMask;
+        }
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Context>
+    {
+        private final Session session;
+        private final VariableAllocator variableAllocator;
+        private final PlanNodeIdAllocator planNodeIdAllocator;
+        private final FunctionAndTypeManager functionAndTypeManager;
+
+        public Rewriter(Session session, VariableAllocator variableAllocator, PlanNodeIdAllocator planNodeIdAllocator, FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.variableAllocator = requireNonNull(variableAllocator, "variableAllocator is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        }
+
+        public static RowExpression ifThenElse(RowExpression... arguments)
+        {
+            return specialForm(SpecialFormExpression.Form.IF, arguments[1].getType(), arguments);
+        }
+
+        @Override
+        public PlanNode visitPlan(PlanNode node, RewriteContext<Context> context)
+        {
+            if (!context.get().isEmpty()) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unexpected plan node between partial and final aggregation");
+            }
+            return context.defaultRewrite(node, context.get());
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Context> context)
+        {
+            PlanNode rewrittenSource = context.rewrite(node.getSource(), context.get());
+            boolean canOptimize = !node.getGroupingKeys().isEmpty() && node.getAggregations().values().stream()
```

Comment:
you should add a comment about why these are the criteria for being able to optimize